base_report_to_printer/models/ir_actions_report.py

```diff
@@ -0,0 +1,161 @@
+    def onchange_printing_printer_id(self):
+        """Reset the tray when the printer is changed"""
+        self.printer_tray_id = False
+
+    @api.model
+    def print_action_for_report_name(self, report_name):
+        """Returns if the action is a direct print or pdf
+
+        Called from js
+        """
+        report = self._get_report_from_name(report_name)
+        if not report:
+            return {}
+        result = report.behaviour()
+        serializable_result = {
+            "action": result["action"],
+            "printer_name": result["printer"].name,
+        }
+        return serializable_result
+
+    def _get_user_default_print_behaviour(self):
+        printer_obj = self.env["printing.printer"]
+        user = self.env.user
+        return dict(
+            action=user.printing_action or "client",
+            printer=user.printing_printer_id or printer_obj.get_default(),
+            tray=str(user.printer_tray_id.system_name)
+            if user.printer_tray_id
+            else False,
+        )
+
+    def _get_report_default_print_behaviour(self):
+        result = {}
+        report_action = self.property_printing_action_id
+        if report_action and report_action.action_type != "user_default":
+            result["action"] = report_action.action_type
+        if self.printing_printer_id:
+            result["printer"] = self.printing_printer_id
+        if self.printer_tray_id:
+            result["tray"] = self.printer_tray_id.system_name
+        return result
+
+    def behaviour(self):
+        self.ensure_one()
+        printing_act_obj = self.env["printing.report.xml.action"]
+
+        result = self._get_user_default_print_behaviour()
+        result.update(self._get_report_default_print_behaviour())
+
+        # Retrieve report-user specific values
+        print_action = printing_act_obj.search(
+            [
+                ("report_id", "=", self.id),
+                ("user_id", "=", self.env.uid),
+                ("action", "!=", "user_default"),
+            ],
+            limit=1,
+        )
+        if print_action:
+            # For some reason design takes report defaults over
+            # False action entries so we must allow for that here
+            result.update({k: v for k, v in print_action.behaviour().items() if v})
+        return result
+
+    def print_document(self, record_ids, data=None):
+        """Print a document, do not return the document file"""
+        document, doc_format = self.with_context(
+            must_skip_send_to_printer=True
+        )._render_qweb_pdf(record_ids, data=data)
+        behaviour = self.behaviour()
+        printer = behaviour.pop("printer", None)
+
+        if not printer:
+            raise exceptions.UserError(_("No printer configured to print this report."))
+        if self.print_report_name:
+            report_file_names = [
+                safe_eval(self.print_report_name, {"object": obj})
```

Comment:
If you import time from odoo.tools.safe_eval you can keep the time :wink: 
[from odoo.tools.safe_eval import safe_eval, time](https://github.com/odoo/odoo/blob/c9980b8f946005904cf4701b71a1c37e7577ea65/addons/web/controllers/main.py#L40)

Reply:
Thanks you very much @etobella for your review